src/pudl/analysis/record_linkage/link_cross_year.py

```diff
@@ -0,0 +1,273 @@
+        for chunk in pairwise_distances_chunked(feature_matrix, metric=config.metric):
+            self.distance_matrix[row_start : row_start + len(chunk), :] = chunk[:, :]
+            self.distance_matrix.flush()
+            row_start += len(chunk)
+
+        # Apply distance penalty to records from the same year
+        year_inds = original_df.groupby("report_year").indices
+        for inds in year_inds.values():
+            matching_year_inds = np.array(np.meshgrid(inds, inds)).T.reshape(-1, 2)
+            self.distance_matrix[
+                matching_year_inds[:, 0], matching_year_inds[:, 1]
+            ] = config.distance_penalty
+
+        np.fill_diagonal(self.distance_matrix, 0)
+        self.distance_matrix.flush()
+
+        # Convert distance matrix to read only memory map
+        self.distance_matrix = np.memmap(
+            filename,
+            dtype="float32",
+            mode="r",
+            shape=(feature_matrix.shape[0], feature_matrix.shape[0]),
+        )
+
+
+def get_cluster_distance_matrix(
+    distance_matrix: np.ndarray, cluster_inds: np.ndarray
+) -> np.ndarray:
+    """Return a distance matrix with only distances within a cluster."""
+    cluster_size = len(cluster_inds)
+    dist_inds = np.array(np.meshgrid(cluster_inds, cluster_inds)).T.reshape(-1, 2)
+    return distance_matrix[dist_inds[:, 0], dist_inds[:, 1]].reshape(
+        (cluster_size, cluster_size)
+    )
+
+
+@njit
+def get_average_distance_matrix(
+    distance_matrix: np.ndarray,
+    cluster_groups: list[list[int]],
+) -> np.ndarray:
+    """Compute average distance between two clusters of records given indices of each cluster."""
+    # Prepare a distance matrix of (n_clusters x n_clusters)
+    # Distance matrix contains average distance between each cluster
+    n_clusters = len(cluster_groups)
+    average_dist_matrix = np.zeros((n_clusters, n_clusters))
+
+    # Heavy nested looping optimized by numba
+    for i, cluster_i in enumerate(cluster_groups):
+        for j, cluster_j in enumerate(cluster_groups[:i]):
+            total_dist = 0
+            for cluster_i_ind in cluster_i:
+                for cluster_j_ind in cluster_j:
+                    total_dist += distance_matrix[cluster_i_ind, cluster_j_ind]
+
+            average_dist = total_dist / (len(cluster_i) + len(cluster_j))
+            average_dist_matrix[i, j] = average_dist
+            average_dist_matrix[j, i] = average_dist
+
+    return average_dist_matrix
+
+
+@op
+def compute_distance_with_year_penalty(
+    config: PenalizeReportYearDistanceConfig,
+    feature_matrix: FeatureMatrix,
+    original_df: pd.DataFrame,
+) -> DistanceMatrix:
+    """Compute a distance matrix and penalize records from the same year."""
+    return DistanceMatrix(feature_matrix.matrix, original_df, config)
+
+
+class DBSCANConfig(Config):
+    """Configuration for DBSCAN step."""
+
+    #: See :class:`sklearn.cluster.DBSCAN` for details.
+    eps: float = 0.1
+    min_samples: int = 2
```

Comment:
Probably out of scope for this PR but it would be nice to potentially set up some experiment tracking infrastructure so we can do some cross validation on model parameters. Harder to do in this context since we have no training data, but in the FERC to EIA process we're running a full grid search that probably shouldn't happen every time, but should instead be documented somewhere else and revisited when data changes?

Reply:
Yeah I think some experiment tracking would be really nice, but I'm not sure when the best time to do that would be. Maybe it's something we could spend some remaining hours in the new year on if time permits, or during the mozilla grant?